src/engine.ml

```diff
@@ -1918,6 +1892,101 @@ let incoming state control_crypto buf =
       m "action %a" Fmt.(option ~none:(any "none") pp_action) act_opt);
   (s', out, List.rev payloads, act_opt)
 
+let new_connection server data =
+  let open Result.Syntax in
+  let protocol = `Tcp in
+  let session =
+    init_session
+      ~my_session_id:(Randomconv.int64 server.server_rng)
+      ~protocol ()
+  in
+  let current_ts = server.server_ts () in
+  let channel = new_channel 0 current_ts in
+  let* t, (control_crypto : control_tls), packet =
+    let tls_auth_or_tls_crypt error =
+      let+ (control_crypto : control_tls) =
+        match
+          (tls_auth server.server_config, tls_crypt server.server_config)
+        with
+        | Ok auth, Error _ -> Ok (`Tls_auth auth)
+        | Error _, Ok crypt -> Ok (`Tls_crypt (crypt, None))
+        | _ -> Error error
+      in
+      ( {
+          config = server.server_config;
+          is_not_taken = server.is_not_taken;
+          control_crypto :> control_crypto;
+          state = Server Server_handshaking;
+          linger = Cstruct.empty;
+          rng = server.server_rng;
+          ts = server.server_ts;
+          now = server.server_now;
+          session;
+          channel;
+          lame_duck = None;
+          last_received = current_ts;
+          last_sent = current_ts;
+        },
+        control_crypto,
+        data )
+    in
+    match Config.find Tls_crypt_v2_server server.server_config with
+    | None ->
+        tls_auth_or_tls_crypt
+          (`Msg "server only supports tls-auth or tls-crypt")
+    | Some (_wrapping_key, true) ->
+        (* TODO: HMAC cookie support *)
+        Error (`Msg "Server does not support hmac cookies (yet)")
+    | Some (wrapping_key, false) -> (
+        match Packet.decode_key_op protocol data with
+        | Error `Tcp_partial ->
+            (* It is unlikely that we don't read a full packet in first try and it is annoying to handle. *)
+            assert false
+        | Error _ as e -> e
+        | Ok (Packet.Hard_reset_client_v2, _key, _received, linger) ->
+            assert (Cstruct.is_empty linger);
+            tls_auth_or_tls_crypt (`Msg "server only supports tls-crypt-v2")
+        | Ok (Packet.Hard_reset_client_v3, _key, received, linger) ->
+            assert (Cstruct.is_empty linger);
+            (* decode and unwrap wKc *)
+            let* actual_packet, wkc =
+              Tls_crypt.Wrapped_key.of_cstruct received
+            in
+            let* tls_crypt, _metadata =
+              Tls_crypt.Wrapped_key.unwrap ~key:wrapping_key wkc
+            in
```

Comment:
you likely know better -- is the wkc encrypted as the other parts of the control packets? according to `handle_client`, it looks like it is encrypted in wrap_and_out calling encrypt_and_out